### PR TITLE
Add support for additional key types and shared secret derivation

### PR DESCRIPTION
**Pull request by chris-dudley**
This PR makes the following changes:
- Modifies the Marshal/Load functions to support any key format supported by OpenSSL.
- Adds support for generating Elliptic Curve key pairs on named curves.
- Adds support for shared secret derivation (e.g., Diffie Hellman) for key types supported by OpenSSL.

I tried to follow existing coding conventions as much as possible. Please let me know if anything is unacceptable and I will work to fix it.

All tests passed with `go test` on the following platforms:
- Linux 3.10.0-327.13.1.el7.x86_64, OpenSSL 1.0.1k-fips, go1.8.3
- Linux 4.13.16-302.fc27.x86_64, OpenSSL 1.1.0g-fips, go1.9.1

--------------

While adding support for other key types for Marshal/Load, I modified the MarshalPKCS1PrivateKey(PEM|DER) functions to use the more general key serialization methods.

This is backwards compatible with the old implementation for RSA keys, but for key types that do not have a PKCS1 format, the keys will be marshaled to PKCS8 instead.

If this is not acceptable, let me know and I will restore the previous implementation and use different names for the new implementation (maybe MarshalTraditionalPrivateKey(PEM|DER) or something).

### CODE REVIEW

**Comment by zeebo** [comment]
I’m going to try to review this, but due to the holidays it might take me some time. Just wanted to let you know :)

Thanks for the contribution!

**Comment by chris-dudley** [comment]
No problem, happy holidays!

**Comment by chris-dudley** [comment]
I just realized that on the machine running OpenSSL 1.0.x, I ran the tests on the wrong branch. There are a few things that break when compiling against that version.

I need to fix these issues, so no worries about reviewing this PR right now.

I'll leave another comment once the issues are resolved.

**Comment by chris-dudley** [comment]
That should fix the incompatibilities with OpenSSL 1.0.x.

File: key.go [hunk 1]
```diff
@@ -117,15 +158,15 @@ func (key *pKey) MarshalPKCS1PrivateKeyPEM() (pem_block []byte,
...
-	}
-	defer C.RSA_free(rsa)
-	if int(C.PEM_write_bio_RSAPrivateKey(bio, rsa, nil, nil, C.int(0), nil,
-		nil)) != 1 {
+
+	// PEM_write_bio_PrivateKey_traditional will use the key-specific PKCS1
+	// format if one is available for that key type, otherwise it will encode
+	// to a PKCS8 key.
+	if int(C.X_PEM_write_bio_PrivateKey_traditional(bio, key.key, nil, nil,
+		C.int(0), nil, nil)) != 1 {
```
Comment (zeebo) [inline key.go]: It's the case that rsa is still written as PKCS1, correct? The naming of this method is now a bit off in the case of it using PKCS8. 😢 Oh well!

Comment (zeebo) [reply key.go]: Oops, just noticed you addressed this in the PR comment.

**Pull request merged**